src/odemis/acq/stitching/_registrar.py

```diff
@@ -427,3 +437,313 @@ def _compute_registration(self, tile, row, col):
+        # requires some modifications to the obtain a proper performance metric:
+        # - lower values are better than higher values --> invert the sign
+        # - the error should never be 0 --> use values in range [100, 200] instead of [0, 100]
+        error = 200 - (ncc + 1) * 50
+
+        overlap = min(numpy.subtract(tile.shape, numpy.abs(exp_shift)))
+        if max(numpy.abs(shift)) > overlap:
+            logging.info("Calculated shift is larger than the overlap size, using expected position "
+                         "instead.")
+            return exp_shift, 100
+
+        return shift_total, error
+
+    def _compute_registration(self, row, col):
+        """
+        Performs registration of the tile at grid position row, col with respect to every
+        available neighbour. The computed shifts and the respective cross-correlation values
+        are stored in self.shifts.
+
+        :param row: (int) row index
+        :param col: (int) col index
+        :updates self.shifts:
+        """
+        tile = self.tiles[row][col]
+        num_cols = len(self.tiles[0])
+        num_rows = len(self.tiles)
+
+        # Neighbouring tiles and shifts
+        nbr_left = self.tiles[row][col - 1] if col > 0 else None
+        nbr_right = self.tiles[row][col + 1] if col < num_cols - 2 else None
+        nbr_top = self.tiles[row - 1][col] if row > 0 else None
+        nbr_bottom = self.tiles[row + 1][col] if row < num_rows - 2 else None
+
+        shift_left = self.shifts['hor'][row][col - 1] if col > 0 else None
+        shift_right = self.shifts['hor'][row][col] if col < num_cols - 2 else None
+        shift_top = self.shifts['ver'][row - 1][col] if row > 0 else None
+        shift_bottom = self.shifts['ver'][row][col] if row < num_rows - 2 else None
+
+        # Calculate the shifts to all adjacent tiles that have not been calculated yet
+        if nbr_left is not None and not shift_left:
+            self.shifts['hor'][row][col - 1] = self._get_shift(nbr_left, tile)
+        if nbr_right is not None and not shift_right:
+            self.shifts['hor'][row][col] = self._get_shift(tile, nbr_right)
+        if nbr_top is not None and not shift_top:
+            self.shifts['ver'][row - 1][col] = self._get_shift(nbr_top, tile)
+        if nbr_bottom is not None and not shift_bottom:
+            self.shifts['ver'][row][col] = self._get_shift(tile, nbr_bottom)
+            
+    def _assemble_mosaic(self):
+        """
+        Performs a global optimization to find the best path through the tile grid using 
+        a minimum spanning tree.
+
+        :returns: (DataArray with shape num_rows x num_cols x 2) registered positions
+        """
+        # Transform shifts and errors to adjacency matrices
+        num_cols = len(self.tiles[0])
+        num_rows = len(self.tiles)
+        errors = numpy.zeros([num_rows * num_cols, num_rows * num_cols])
```

Comment:
Please, could you use a tuple to indicate the shape. It might be me, but I find it clearer.